url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/rtree/RTree.java

```diff
@@ -0,0 +1,1674 @@
+import spatialindex.io.IStorageManager;
+import spatialindex.io.InvalidPageException;
+import spatialindex.io.RandomEvictionsBuffer;
+import spatialindex.setting.PropertySet;
+import spatialindex.spatialindex.SpatialIndex;
+import utility.Compare.DataOfPoint;
+import utility.geo.DataOfLine;
+import utility.security.DataIO;
+import utility.security.SecurityUtility;
+
+public class RTree implements ISpatialIndex {
+	
+	public RandomAccessFile rafPoint = null;
+	public RandomAccessFile rafLine = null;
+	public HashMap<Long, long[]> tOfPoint = null;
+	public HashMap<Long, long[]> tOfLine = null;
+	public HashMap<Long, DataOfPoint> buffer_Points = new HashMap<Long, DataOfPoint>();
+	public HashMap<Long, DataOfLine> buffer_Lines = new HashMap<Long, DataOfLine>();
+	public String destFileNamePoint, destFileNameLine;
+	IShape queryRegion = null;
+
+	RWLock m_rwLock;
+
+	IStorageManager m_pStorageManager;
+
+	int m_rootID;
+	int m_headerID;
+
+	/**
+	 * indicates different trees, see SpatialIndex.RtreeVariantLinear
+	 */
+	int m_treeVariant;
+
+	double m_fillFactor;
+
+	int m_indexCapacity;
+
+	int m_leafCapacity;
+
+	int m_nearMinimumOverlapFactor;
+	// The R*-Tree 'p' constant, for calculating nearly minimum overlap cost.
+	// [Beckmann, Kriegel, Schneider, Seeger 'The R*-tree: An efficient and Robust Access Method
+	// for Points and Rectangles, Section 4.1]
+
+	double m_splitDistributionFactor;
+	// The R*-Tree 'm' constant, for calculating spliting distributions.
+	// [Beckmann, Kriegel, Schneider, Seeger 'The R*-tree: An efficient and Robust Access Method
+	// for Points and Rectangles, Section 4.2]
+
+	double m_reinsertFactor;
+	// The R*-Tree 'p' constant, for removing entries at reinserts.
+	// [Beckmann, Kriegel, Schneider, Seeger 'The R*-tree: An efficient and Robust Access Method
+	//  for Points and Rectangles, Section 4.3]
+
+	int m_dimension;
+
+	private int update_node_size;
+	Region m_infiniteRegion;
+
+	Statistics m_stats;
+
+	ArrayList<INodeCommand> m_writeNodeCommands = new ArrayList<INodeCommand>();
+	ArrayList<INodeCommand> m_readNodeCommands = new ArrayList<INodeCommand>();
+	ArrayList<INodeCommand> m_deleteNodeCommands = new ArrayList<INodeCommand>();
+
+	/**
+	 * @return the update_node_size
+	 */
+	public int getUpdate_node_size() {
+		return update_node_size;
+	}
+
+	/**
+	 * @param update_node_size the update_node_size to set
+	 */
+	public void setUpdate_node_size(int update_node_size) {
+		this.update_node_size = update_node_size;
+	}
+
+	public void setDestFileNamePoint(String name){
+		destFileNamePoint = name;
+	}
+	
+	public void setDestFileNameLine(String name){
+		destFileNameLine = name;
+	}
+	
+	public static RTree createRTree(String[] args) {
+		try {
+			if (args.length != 6) {
+				println("Usage: RTreeLoad input_file tree_file capacity query_type [intersection | 10NN].");
+				System.exit(-1);
+			}
+
+			LineNumberReader lr = null;
+
+			try {
+				lr = new LineNumberReader(new FileReader(args[0]));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *lr*. Currently, there are execution paths that do not contain closure statements, for example, when an else condition at line 342 evaluates to true. Ensure *lr* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).